blueprints/to-locks.md

```diff
@@ -0,0 +1,281 @@
+`soft` vs `hard` locks, and forbid a user from snapping/ queueing a CDN if another user possesses a lock on that CDN.
+The following endpoints will need to handle the locks logic:
+- `/cachegroups/{{ID}}/queue_update`
+- `/cdns/{{ID}}/queue_update`
+- `/servers/{{hostname}}/queue_update`
+- `/snapshot`
+- `/topologies/{{name}}/queue_update`
+
+In addition to these, all `PUT`, `POST`, `DELETE` endpoints that scope to a CDN will have to add in the logic to check if there 
+is a "hard" lock by some other user, before a user can modify data.
+Basically, every endpoint that affects data related to a CDN directly (for example, `profiles`), and every endpoint that affects data 
+related to a CDN indirectly (for example, `parameters`, `cachegroups`) will need to be updated to perform the above mentioned check.
+
+#### REST API Impact
+
+The following is the JSON representation of a `CDN_Lock` object:
+```JSON
+{
+  "username": "foo",
+  "cdn": "cdn1",
+  "message": "snapping cdn",
+  "soft": false,
+  "lastUpdated": "2021-05-10T16:03:34-06:00"
+}
+```
+
+The following table describes the top level `CDN_Lock` object:
+
+| field       | type                        | optionality | description                                                                                                                  |
+| ----------- | --------------------------- | ----------- | -----------------------------------------------------------------------------------------------------------------------------|
+| username    | string                      | required    | the user name of the user that wants to acquire a lock on the CDN                                                            |
+| cdn         | string                      | required    | the name of the CDN on which the lcok needs to be acquired                                                                   |
+| message     | string                      | optional    | the message stating a reason behind locking the CDN                                                                          |
+| soft        | boolean                     | required    | whether or not this is a shared lock, meaning if a user has the lock, whether or not other users can make changes to the CDN |
+| lastUpdated | time                        | required    | the last time this lock was updated                                                                                          |
+
+**API constraints:**
+- a CDN can have only one `hard` or `soft` lock at a time
+- a user can snap/ queue a CDN only if they are the one holding the lock (of either kind) on the `cdn`. Alternatively, if no one has the lock on the `cdn`, anyone can snap/ queue/ make changes to the CDN
+- a user can delete their `hard` or `soft` lock whenever they want
+- a user with at least `all-read` and `all-write` capabilities can delete the locks of other users on any CDN. The `all-write` capability will be modified
```

Comment:
this is fine but permissions don't work that way currently and maybe not for some time so you need to pin each action to a role.

GET locks/lock = read/only role
POST locks = operations+ role?
DELETE lock = operations+ role and only your own lock unless admin role?

Reply:
Yeah, sorry, I should've been more clear. It should still specify the privilege levels, since that's what we actually use today